### PR TITLE
Threat vault v2

### PR DESCRIPTION
**Pull request by israelpoli**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-460

## Description
New integration of ThreatVault.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ShirleyDenkberg** [comment]
@israelpoli  There are several comments in the yml file (such as: ## checks, ## not done, ## not done, I need add the outputs). Please address/remove these comments.
@esharf @bakatzir @GuyAfik Doc review completed.

**Review by dorschw — COMMENTED** [review]
Well done!

File: Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_application.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
+{
+    "id": "incident_threatvaultbypaloaltonetworksapplicationsnew",
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_application.json:2]:
```suggestion
    "id": "incident_threatvaultbypaloaltonetworksapplications",
```
here and in other places

File: Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_application.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    "associatedTypes": [
+        "Threat Vault Release notes"
+    ],
+    "associatedToAll": false,
+    "unmapped": false,
+    "unsearchable": true,
+    "caseInsensitive": true,
+    "columns": [
+        {
+            "key": "columnheader1",
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_application.json:30]: what's this?

File: Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_content_version.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
+    "id": "incident_threatvaultbypaloaltonetworkscontentversion",
+    "version": -1,
+    "modified": "2022-12-14T08:45:16.703899923Z",
+    "name": "Threat Vault by Palo Alto Networks Content Version",
+    "ownerOnly": false,
+    "cliName": "threatvaultbypaloaltonetworkscontentversion",
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_content_version.json:7]: do we have a built in field for this?

File: Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_file_type.json [hunk 4]
```diff
@@ -0,0 +1,77 @@
+{
+    "id": "incident_threatvaultbypaloaltonetworksfiletypenew",
+    "version": -1,
+    "modified": "2022-12-13T15:18:28.981712265Z",
+    "name": "Threat Vault by Palo Alto Networks file type (new)",
+    "ownerOnly": false,
+    "cliName": "threatvaultbypaloaltonetworksfiletypenew",
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_file_type.json:7]: do we have a built-in field for this?

File: Packs/PaloAltoNetworks_Threat_Vault/Scripts/MarkdownRepresentationThreatVaultIncident/MarkdownRepresentationThreatVaultIncident.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            data.update({KEYS_INCIDENT_FIELDS[key]: field_content})
+
+    return data
+
+
+def json_to_md(incident_fields: dict) -> str:
+
+    md = ''
+
+    for incident_field in incident_fields.keys():
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Scripts/MarkdownRepresentationThreatVaultIncident/MarkdownRepresentationThreatVaultIncident.py:31]:
```suggestion
    for incident_field in incident_fields:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Scripts/MarkdownRepresentationThreatVaultIncident/MarkdownRepresentationThreatVaultIncident.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+commonfields:
+  id: MarkdownRepresentationThreatVaultIncident
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Scripts/MarkdownRepresentationThreatVaultIncident/MarkdownRepresentationThreatVaultIncident.yml:2]: let's rename it SetThreatVaultIncidentMarkdownRepresentation

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 7]
```diff
@@ -0,0 +1,790 @@
...
+
+    if 'antivirus' in response['data']:
+        if expanded:
+            antiviruses = response.get('data', {}).get('antivirus', [])
+        else:
+            antiviruses = [response.get('data', {}).get('antivirus', ([], ...))[0]]
+
+        for antivirus in antiviruses:
+            table_for_md = resp_to_hr(response=antivirus, type_='antivirus')
+            readable_output = tableToMarkdown(name=f"{antivirus.get('id')} antivirus reputation:", t=table_for_md,
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:278]:
```suggestion
            readable_output = tableToMarkdown(name=f"Antivirus Reputation: {antivirus.get('id')}", t=table_for_md,
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 8]
```diff
@@ -0,0 +1,790 @@
...
+            )
+
+    if 'spyware' in response['data']:
+        if expanded:
+            spywares = response.get('data', {}).get('spyware', [])
+        else:
+            spywares = [response.get('data', {}).get('spyware', ([], ...))[0]]
+        for spyware in spywares:
+            table_for_md = resp_to_hr(response=spyware, type_='spyware')
+            readable_output = tableToMarkdown(name=f"{spyware.get('id')} spyware reputation:", t=table_for_md,
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:296]:
```suggestion
            readable_output = tableToMarkdown(name=f"Spyware Reputation: {spyware.get('id')}", t=table_for_md,
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+
+    if 'vulnerability' in response['data']:
+        if expanded:
+            vulnerabilities = response.get('data', {}).get('vulnerability', [])
+        else:
+            vulnerabilities = [response.get('data', {}).get('vulnerability', ([], ...))[0]]
+        for vulnerability in vulnerabilities:
+            table_for_md = resp_to_hr(response=vulnerability, type_='vulnerability')
+            readable_output = tableToMarkdown(name=f"{vulnerability.get('id')} vulnerability reputation:", t=table_for_md,
+                                              removeNull=True)
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:315]:
```suggestion
            readable_output = tableToMarkdown(name=f"Vulnerability Reputation: {vulnerability.get('id')}", t=table_for_md,
                                              removeNull=True)
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 10]
```diff
@@ -0,0 +1,790 @@
...
+
+    if 'fileformat' in response['data']:
+        if expanded:
+            filesformat = response.get('data', {}).get('fileformat', [])
+        else:
+            filesformat = [response.get('data', {}).get('fileformat', ([], ...))[0]]
+        for fileformat in filesformat:
+            table_for_md = resp_to_hr(response=fileformat, type_='fileformat')
+            readable_output = tableToMarkdown(name=f"{fileformat.get('id')} fileformat reputation:", t=table_for_md,
+                                              removeNull=True)
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:333]:
```suggestion
            readable_output = tableToMarkdown(name=f"Fileformat Reputation: {fileformat.get('id')}", t=table_for_md,
                                              removeNull=True)
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 11]
```diff
@@ -0,0 +1,790 @@
...
+            command_results_list.append(
+                CommandResults(
+                    outputs_prefix='ThreatVault.Fileformat',
+                    outputs_key_field='id',
+                    outputs=response.get('data', {}).get('fileformat', []),
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:338]: let's reconsider this

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 12]
```diff
@@ -0,0 +1,790 @@
...
+            file = Common.File(
+                sha256=file_info.get('sha256'),
+                md5=file_info.get('md5'),
+                sha1=file_info.get('sha1'),
+                dbot_score=dbot_score
+            )
+
+            table_for_md = resp_to_hr(response=file_info, type_='file', expanded=args.get('expanded', False))
+
+            readable_output = tableToMarkdown(name=f"Hash {_hash} antivirus reputation:", t=table_for_md,
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:403]:
```suggestion
            readable_output = tableToMarkdown(name=f"Antivirus Reputation for hash: {_hash}", t=table_for_md,
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 13]
```diff
@@ -0,0 +1,790 @@
...
+            vulnerability = response.get('data', {}).get('vulnerability', ([], ...))[0]
+            _cve = Common.CVE(
+                id=vulnerability.get('cve')[0],
+                cvss=vulnerability.get('severity'),
+                published=vulnerability.get('ori_release_time'),
+                modified=vulnerability.get('latest_release_time'),
+                description=vulnerability.get('description'),
+            )
+            table_for_md = resp_to_hr(response=vulnerability, type_='cve')
+            readable_output = tableToMarkdown(name=f"CVE {cve} vulnerability reputation:", t=table_for_md,
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:445]:
```suggestion
            readable_output = tableToMarkdown(name=f"CVE Vulnerability Reputation: {cve}", t=table_for_md,
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+
+def threat_signature_get_command(client: Client, args: Dict) -> List[CommandResults]:
+
+    args['file'] = args.get('sha256', '')
+    if md5 := args.get('md5'):
+        args['file'] += f",{md5}" if args['file'] else md5
+    args['expanded'] = True
+    ids = argToList(args.get('signature_id'))
+
+    if not ids and not args['file']:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:467]:
```suggestion
    if not any((ids, args['file'])):
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+        if response:
+            command_results_list.extend(parse_resp_by_type(response=response))
+
+    return command_results_list
+
+
+def release_note_get_command(client: Client, args: Dict) -> CommandResults:
+
+    if not args.get('version'):
+        raise ValueError('The version argument is required')  # UT
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:503]:
```suggestion
        raise ValueError('The version argument is required') 
```

`# UT` ?

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 16]
```diff
@@ -0,0 +1,790 @@
...
+        response = client.release_notes_get_request('content', version)
+    except Exception as err:
+        if err.res.status_code == 404:  # type:ignore # pylint: disable=E1101
+            return CommandResults(
+                readable_output=f'{version} release note not found.'
+            )
+        else:
+            raise
+
+    data = response.get('data', ([], ...))[0]
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:516]:
```suggestion
    data = response.get('data', ([],))[0]
```

Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:516]: unsafe accessor

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 17]
```diff
@@ -0,0 +1,790 @@
...
+    if not args.get('version'):
+        raise ValueError('The version argument is required')  # UT
+
+    version = args['version']
+    try:
+        response = client.release_notes_get_request('content', version)
+    except Exception as err:
+        if err.res.status_code == 404:  # type:ignore # pylint: disable=E1101
+            return CommandResults(
+                readable_output=f'{version} release note not found.'
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:511]:
```suggestion
                readable_output=f'release note {version} was not found.'
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 18]
```diff
@@ -0,0 +1,790 @@
...
+def fetch_incidents(client: Client, args: dict) -> List:
+    '''
+    Retrieving release notes that contain all the information about vulnerabilities, antivirus, spyware, and more.
+    '''
+
+    last_run = demisto.getLastRun()
+    first_fetch = args.get('first_fetch', '3 Days')
+    if not last_run.get('scound_fetch'):
+        if (first_fetch.strip().split(' ')[1].lower() not in
+           ['days', 'month', 'months', 'year', 'years']):  # only these are allowed
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:678]:
```suggestion
           frozenset(('days', 'month', 'months', 'year', 'years')):  # only these are allowed
```

File: Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_content_version.json (hunk 3 above)
Comment (israelpoli) [reply Packs/PaloAltoNetworks_Threat_Vault/IncidentFields/incidentfield-ThreatVault_content_version.json:7]: I didn't find one

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 19]
```diff
@@ -0,0 +1,916 @@
...
+    )
+
+    for _hash in hashes:
+        type_hash = get_hash_type(_hash)
+        try:
+            response = client.antivirus_signature_get_request(
+                arg=type_hash, value=_hash
+            )
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:438]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 20]
```diff
@@ -0,0 +1,916 @@
...
+def cve_command(client: Client, args: Dict) -> List[CommandResults]:
+
+    cves = argToList(args.get("cve"))
+    command_results_list: List[CommandResults] = []
+
+    for cve in cves:
+        try:
+            response = client.antivirus_signature_get_request(arg="cve", value=cve)
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:508]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 21]
```diff
@@ -0,0 +1,916 @@
...
+
+    if args["file"]:
+        command_results_list.extend(file_command(client=client, args=args))
+        return command_results_list
+
+    for _id in ids:
+        try:
+            response = client.antivirus_signature_get_request(arg="id", value=_id)
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:572]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 22]
```diff
@@ -0,0 +1,916 @@
...
+def release_note_get_command(client: Client, args: Dict) -> CommandResults:
+
+    if not args.get("version"):
+        raise ValueError("The version argument is required")
+
+    version = args["version"]
+    try:
+        response = client.release_notes_get_request("content", version)
+    except DemistoException as err:
+        if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:596]:
```suggestion
        if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 23]
```diff
@@ -0,0 +1,916 @@
...
+    command_results_list: List[CommandResults] = []
+
+    if ids or names:
+        type_ = "id" if ids else "name"
+        try:
+            response = client.threat_batch_search_request(
+                arg=type_, value=ids if ids else names, type_=threat_type
+            )
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:639]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 24]
```diff
@@ -0,0 +1,916 @@
...
+        dbot_reliability = DBotScoreReliability.get_dbot_score_reliability_from_str(
+            client.reliability
+        )
+        type_ = "md5" if md5 else "sha256"
+        try:
+            response = client.threat_batch_search_request(
+                arg=type_, value=md5 or sha256, type_=threat_type
+            )
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:661]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 25]
```diff
@@ -0,0 +1,916 @@
...
+        offset=offset,
+        limit=limit,
+    )
+
+    command_results_list: List[CommandResults] = []
+
+    try:
+        response = client.threat_search_request(args=query)
+    except DemistoException as err:
+        if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:760]:
```suggestion
        if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py [hunk 26]
```diff
@@ -0,0 +1,916 @@
...
+    incidents: List[dict] = []
+    while current <= now:
+        try:
+            # Bringing the daily date for the first api call
+            demisto.debug(f"Time for request fetch-incidents -> {current}")
+            response = client.threat_search_request(
+                {"releaseDate": current.strftime("%Y-%m-%d")}
+            )
+        except DemistoException as err:
+            if err.res.status_code == 404:
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2.py:807]:
```suggestion
            if err.res and err.res.status_code == 404:
```

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py [hunk 27]
```diff
@@ -0,0 +1,923 @@
...
+        "threat_batch_search_request",
+        side_effect=DemistoException(message="test", res=MockException(404)),
+    )
+
+    results = cmd(client, demisto_args)
+
+    if expected_indicator:
+        assert isinstance(results[0].indicator, expected_indicator)
+    if isinstance(results, list):
+        assert results[0].readable_output == expected_readable_output
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py:194]: please add an assertion for the length of `results`

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py [hunk 28]
```diff
@@ -0,0 +1,923 @@
...
+@pytest.mark.parametrize(
+    "page, page_size, limit, expected_result",
+    [(5, 100, None, (500, 100)), (None, None, 100, (0, 100))],
+)
+def test_pagination(page, page_size, limit, expected_result):
+
+    results = pagination(page, page_size, limit)
+    assert results[0] == expected_result[0]
+    assert results[1] == expected_result[1]
+
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py:208]: please add an assertion for the length of `results`

File: Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py [hunk 29]
```diff
@@ -0,0 +1,923 @@
...
+        0,
+    ),
+]
+
+
+@pytest.mark.parametrize("resp, type_, expanded, expected", RESP_TO_HR_ARGS)
+def test_resp_to_hr(resp, type_, expanded, expected):
+
+    result = resp_to_hr(resp, type_, expanded)
+    assert len(result.keys()) == expected
```
Comment (dorschw) [inline Packs/PaloAltoNetworks_Threat_Vault/Integrations/ThreatVaultv2/ThreatVaultv2_test.py:390]: please a test for the output content

**Pull request merged**